### DISCUSSION TITLE
Some plugin API suggestions

### DISCUSSION THREAD

**Discussion by zakk4223** [hyprwm/Hyprland#3868] [General]
This is mostly focused on layout plugins, because I can't seem to stop writing layout plugins...

(Most of this stuff is pretty simple, and I'll gladly do the work/PR. But I'd rather fight you before I write code)

add/remove layout events

plugin API call to retrieve layouts by name. Or maybe just make m_vLayouts public so it can be enumerated/searched etc.

hyprctl command to retrieve info on all the loaded layouts. 
Beyond layout info, maybe a way for plugins to register hyprctl commands and provide the output for them?

I'd also like some way to allow users to have something like a windowrule that sets  properties on a window that third party layouts can use to modify the layout's behavior. I guess each layout plugin could implement a keyword to do this, but it seems annoying to have to duplicate the window matching code in every plugin. Also there would be no way to share that 'rule' between layouts. (I have....ideas)
I feel like this is just asking for arbitrary user 'tags' on windows and I'm not sure how I feel about that...

Last thing, but not really plugin related. I'm already droning on about layouts so this is as good a place as any.

It's impossible to create a tiling layout where the tiled windows overlap. Even if you manipulate the compositor's window list to put them in the 'correct' order. The problem is the window rendering code renders through that list in one order, and all the functions that deal with mouse-based focus search in the *opposite* order. So mouseover focus pulls windows from below the visible one. 

I got around this by hooking most of the VectorToWindow* functions to also search the non-float window list using std::reverse (just like the floats are searched) It seems to work but I don't know if there's a better way or if I'm subtly breaking something important.

Yes overlapping tiled windows are kinda dumb, but such layouts exist. I'm sure at least 2 people would even use them.

**Comment by vaxerski** [comment]
> add/remove layout events

layoutmsg exists

> plugin API call to retrieve layouts by name. Or maybe just make m_vLayouts public so it can be enumerated/searched etc.

sounds ok, but you can just `#define private public`

> hyprctl command to retrieve info on all the loaded layouts.

sounds ok

> Beyond layout info, maybe a way for plugins to register hyprctl commands and provide the output for them?

sounds ok

> I'd also like some way to allow users to have something like a windowrule that sets properties on a window that third party layouts can use to modify the layout's behavior. I guess each layout plugin could implement a keyword to do this, but it seems annoying to have to duplicate the window matching code in every plugin. Also there would be no way to share that 'rule' between layouts. (I have....ideas)
I feel like this is just asking for arbitrary user 'tags' on windows and I'm not sure how I feel about that...

Yeeaaaa, might be worth a debate

> It's impossible to create a tiling layout where the tiled windows overlap. Even if you manipulate the compositor's window list to put them in the 'correct' order. The problem is the window rendering code renders through that list in one order, and all the functions that deal with mouse-based focus search in the opposite order. So mouseover focus pulls windows from below the visible one.

I have had "rewrite the fucking window focus functions for fucks sake" on my to-do list for a while. Maybe I finally should.

**Reply by zakk4223** [reply]
> > add/remove layout events
> 
> layoutmsg exists

Sorry bad wording. I want hookable events when other plugins add/remove layouts. Mostly so I can clean up if a layout gets removed due to plugin unload or whatever. Although maybe it's a better idea to have the layout manager store those as shared_ptr and then hold weak_ptr references to it? 

> 
> > plugin API call to retrieve layouts by name. Or maybe just make m_vLayouts public so it can be enumerated/searched etc.
> 
> sounds ok, but you can just `#define private public`

I tried that and it went poorly. Maybe I gotta fiddle with the order I include stuff...

**Reply by vaxerski** [reply]
`#include` includes.hpp first and then do the define trick, should work. Remember to undef after including hyprland headers.

**Reply by zakk4223** [reply]
for future people showing up here via searches: the #define private public thing does work, but you may have to include some C++ system  includes before you do it. I had to `#include <any>` and `#include <sstream>`

**Reply by vaxerski** [reply]
or redirect them to https://github.com/hyprwm/hyprland-plugins/blob/cc8f72a32e0936d94fd36e78d8ce417b4c6ec436/hyprwinwrap/main.cpp#L9-L14 :P

**Comment by zakk4223** [comment]
Hey I have more bad ideas/questions:

How can the workspace rule layoutopts be opened up to plugins? Plugin API to register a layoutopt namespace ("layoutopt:nstack:" and a callback to encode the value(s) to std::any? Let users just crap whatever into that field and plugins/layouts grab whatever they want and convert as needed? (that seems terrible tho). 

Should everything in that map be namespaced to the layout using it? Like even the existing layoutopt:orientation becomes layoutopt:master:orientation?

**Reply by vaxerski** [reply]
I will open it more, the impl by whoever made the MR was a bit rushed. Ping me tomorrow.

**Reply by zakk4223** [reply]
DO IT

**Reply by vaxerski** [reply]
sorry for the delay, ea7569d7e0941d19f5f469a5fbb79bc0fa62b935

**Reply by zakk4223** [reply]
That'll work. I can just uniquely name my config opts based on the layout name.

There's no way to have a workspace rule that matches 'any workspace on monitor <X>' is there?

**Reply by vaxerski** [reply]
not yet
